### PR TITLE
Review information structure and mark some fields as optional

### LINKED ISSUE #89: Review information structure and mark some fields as optional

### PR DESCRIPTION
**Pull request by mecies**
### Description
In this PR information structure of the app was reviewed and some fields were marked as optional.

You can preview the deploy under [this link](https://5f85ee1a01461c000891bcf6--coderscard.netlify.app/)

As you can see the `testimonials` in the `about` page are hidden` because they are empty in the content file.
Also, the blog page is completely optional and temporarily I added a redirect from the page when it's empty but later we want to hide it in the navbar to. This is outside of this PRs scope so I didn't adjust the navbar.
For the optional fields, I adjusted their types and added type checks in code where they are used.
I also marked the phone number as optional for privacy reasons and social media links too because not everyone uses every social media we include. In the about page, I also marked technologies, tools, and other skills as optional.
On the portfolio page, I marked link to code, link to mockups, and link to app as optional because not everyone has those included in their projects (considering the fact that we want to encourage not only a developer but also, let's say designers to use the template)

Comments and suggestions are more than welcome 😄 

### TLDR

**Changes**:
- About page: technologies, tools, other skills, testimonials, marked as optional
- Blog page: entire page is optional
- Developer profile: phone number and social media are optional
- Portfolio page: link to code, link to mockups, link to an app, are marked as optional

**Preview**: [link](https://5f85ee1a01461c000891bcf6--coderscard.netlify.app/)

Closes #89

### CODE REVIEW

File: src/pages/blog/index.tsx [hunk 1]
```diff
@@ -160,32 +166,33 @@ const BlogPage: FC<{ data: BlogGQL }> = ({ data }) => {
...
-                    tagtitle={blogPost.blogLabel}
-                    title={blogPost.blogTitle}
-                    type={componentType}
-                    handleClose={() => handleCloseBlogpost()}
-                    handlePrev={() => handlePrevBlogpost(index)}
-                    handleNext={() => handleNextBlogpost(index)}
-                  />
-                </div>
-              ))}
+              {blogData.blogPost !== undefined &&
```
Comment (KonradSzwarc) [inline src/pages/blog/index.tsx]: https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx

File: src/pages/blog/index.tsx [hunk 2]
```diff
@@ -160,32 +166,33 @@ const BlogPage: FC<{ data: BlogGQL }> = ({ data }) => {
...
+                      title={blogPost.blogTitle}
+                      date={dayjs(blogPost.publishDate).format('DD MMMM YYYY')}
+                      onClick={() => setSelectedBlogpost(index)}
+                    />
+                    <BlogPostDialog
+                      contentheader={blogPost.blogDescription}
+                      contentmain={blogPost.blogBody}
+                      imgurl={blogPost.blogImage.publicURL}
+                      isOpen={index === selectedBlogpost}
+                      subtitle={dayjs(blogPost.publishDate).format('DD MMMM YYYY')}
```
Comment (KonradSzwarc) [inline src/pages/blog/index.tsx]: Move date format string to a separate constant

File: src/pages/index.tsx [hunk 3]
```diff
@@ -175,50 +175,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
-                />
-              ))}
+              {aboutData.socialMedia.technologies && aboutData.socialMedia.technologies.length > 0 && (
+                <>
+                  <Typography variant="h6" className={classes.skillsHeader}>
+                    Technologies
+                  </Typography>
+                  <Box className={classes.skills}>
+                    {aboutData.socialMedia.technologies?.map((item) => (
+                      <Skill key={`${item.technologyName}`} level={item.technologyValue}>
```
Comment (KonradSzwarc) [inline src/pages/index.tsx]: `key={item.technologyName}` could be enough

File: src/pages/index.tsx [hunk 4]
```diff
@@ -175,50 +175,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
+                </>
+              )}
+              {aboutData.socialMedia.tools && aboutData.socialMedia.tools.length > 0 && (
+                <>
+                  <Typography variant="h6" className={classes.skillsHeader}>
+                    Tools
+                  </Typography>
+                  <Box className={classes.skills}>
+                    {aboutData.socialMedia.tools?.map((item) => (
+                      <Skill key={`${item.toolName}`} level={item.toolValue}>
```
Comment (KonradSzwarc) [inline src/pages/index.tsx]: Use just `key={item.toolName}`

File: src/pages/index.tsx [hunk 5]
```diff
@@ -175,50 +175,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
+                </>
+              )}
+              {aboutData.socialMedia.otherSkills && aboutData.socialMedia.otherSkills.length > 0 && (
+                <>
+                  <Typography variant="h6" className={classes.skillsHeader}>
+                    Other skills
+                  </Typography>
+                  <Box className={classes.skills}>
+                    {aboutData.socialMedia.otherSkills?.map((item) => (
+                      <Skill key={`${item.otherSkillName}`} level={item.otherSkillValue}>
```
Comment (KonradSzwarc) [inline src/pages/index.tsx]: `key={item.otherSkillName}`

File: src/pages/index.tsx [hunk 6]
```diff
@@ -175,50 +175,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
+                </>
+              )}
             </Box>
+            {aboutData.testimonials && aboutData.testimonials?.length > 0 && (
+              <>
+                <SectionTitle className={classes.title}>Testimonials</SectionTitle>
+                <Box className={classes.testimonials}>
+                  {aboutData.testimonials.map((item) => (
+                    <Testimonial
+                      key={`${item.testimonialName}`}
```
Comment (KonradSzwarc) [inline src/pages/index.tsx]: `key={item.testimonialName}`

File: src/pages/portfolio/index.tsx [hunk 7]
```diff
@@ -182,9 +182,9 @@ const PortfolioPage: FC<{ data: ProjectGQL }> = ({ data }) => {
                     contentMainRole={project.projectRole}
                     contentHeader={project.projectPreviewNote}
                     tagtitle={project.projectLabel}
-                    mockupsUrl={project.projectMockups}
-                    projectUrl={project.projectApp}
-                    codeUrl={project.projectCode}
+                    mockupsUrl={project.projectMockups || ''}
```
Comment (KonradSzwarc) [inline src/pages/portfolio/index.tsx]: Use `??`

**Review by KonradSzwarc — CHANGES_REQUESTED** [review]
A few suggestions + if possible, apply suggestions from the `GitHub Actions / Run ESLint`

File: src/pages/blog/index.tsx (hunk 1 above)
Comment (mecies) [reply src/pages/blog/index.tsx]: Nice read, thanks, updated to ternary because I'm not a fan of `ifs` just like the author 😄

File: src/pages/blog/index.tsx (hunk 2 above)
Comment (mecies) [reply src/pages/blog/index.tsx]: Good idea, done 👍

File: src/pages/index.tsx (hunk 3 above)
Comment (mecies) [reply src/pages/index.tsx]: Good catch 1/4

File: src/pages/index.tsx (hunk 4 above)
Comment (mecies) [reply src/pages/index.tsx]: Good catch 2/4

File: src/pages/index.tsx (hunk 5 above)
Comment (mecies) [reply src/pages/index.tsx]: Good catch 3/4

File: src/pages/index.tsx (hunk 6 above)
Comment (mecies) [reply src/pages/index.tsx]: Good catch 4/4 😄

File: src/pages/portfolio/index.tsx (hunk 7 above)
Comment (mecies) [reply src/pages/portfolio/index.tsx]: Good idea, changed 😄

File: src/pages/blog/index.tsx [hunk 8]
```diff
@@ -122,25 +122,32 @@ const BlogPage: FC<{ data: BlogGQL }> = ({ data }) => {
   const { componentType, isDesktop } = useComponentType();
   const developerProfile = useDeveloperProfile();
   const [selectedBlogpost, setSelectedBlogpost] = useState(-1);
+  const dateFormat = 'DD MMMM YYYY';
+
+  if (typeof window !== `undefined`) {
```
Comment (KonradSzwarc) [inline src/pages/blog/index.tsx]: Why nested if?

File: src/pages/index.tsx [hunk 9]
```diff
@@ -175,50 +175,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
-                <Testimonial
-                  key={`${item.testimonialName}`}
-                  isMobile={isMobile}
-                  image={item.testimonialImage.publicURL}
-                  description={item.testimonialText}
-                  labelBold={item.testimonialName}
-                  label={item.testimonialJob}
-                />
-              ))}
+              {aboutData.socialMedia.technologies && aboutData.socialMedia.technologies.length > 0 && (
```
Comment (KonradSzwarc) [inline src/pages/index.tsx]: Move it to a separate variable to describe its meaning (e.g. `hasTechnologies`)

File: src/pages/index.tsx [hunk 10]
```diff
@@ -175,50 +175,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
+                  <Box className={classes.skills}>
+                    {aboutData.socialMedia.technologies?.map((item) => (
+                      <Skill key={item.technologyName} level={item.technologyValue}>
+                        {item.technologyName}
+                      </Skill>
+                    ))}
+                  </Box>
+                </>
+              )}
+              {aboutData.socialMedia.tools && aboutData.socialMedia.tools.length > 0 && (
```
Comment (KonradSzwarc) [inline src/pages/index.tsx]: Move it to a separate variable to describe its meaning

File: src/pages/index.tsx [hunk 11]
```diff
@@ -175,50 +175,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
+                  <Box className={classes.skills}>
+                    {aboutData.socialMedia.tools?.map((item) => (
+                      <Skill key={item.toolName} level={item.toolValue}>
+                        {item.toolName}
+                      </Skill>
+                    ))}
+                  </Box>
+                </>
+              )}
+              {aboutData.socialMedia.otherSkills && aboutData.socialMedia.otherSkills.length > 0 && (
```
Comment (KonradSzwarc) [inline src/pages/index.tsx]: Move it to a separate variable to describe its meaning

File: src/pages/index.tsx [hunk 12]
```diff
@@ -175,50 +175,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
+                    {aboutData.socialMedia.otherSkills?.map((item) => (
+                      <Skill key={item.otherSkillName} level={item.otherSkillValue}>
+                        {item.otherSkillName}
+                      </Skill>
+                    ))}
+                  </Box>
+                </>
+              )}
             </Box>
+            {aboutData.testimonials && aboutData.testimonials?.length > 0 && (
```
Comment (KonradSzwarc) [inline src/pages/index.tsx]: Move it to a separate variable to describe its meaning

**Review by KonradSzwarc — CHANGES_REQUESTED** [review]

File: src/pages/blog/index.tsx (hunk 8 above)
Comment (mecies) [reply src/pages/blog/index.tsx]: No idea, I was fixing it chaotically after the production build failed,  changed to one `if` 👌 
Fun fact, Gatsby production build is SSR so we need to check for `window` if we want to reference it, for example, the `navigate` function under the hood uses it and it was breaking the build.

File: src/pages/index.tsx (hunk 9 above)
Comment (mecies) [reply src/pages/index.tsx]: Alrighty 1/4 😄

File: src/pages/index.tsx (hunk 10 above)
Comment (mecies) [reply src/pages/index.tsx]: Alrighty 2/4 😄

File: src/pages/index.tsx (hunk 11 above)
Comment (mecies) [reply src/pages/index.tsx]: Alrighty 3/4 😄

File: src/pages/index.tsx (hunk 12 above)
Comment (mecies) [reply src/pages/index.tsx]: Alrighty 4/4 😄

**Review by KonradSzwarc — APPROVED** [review]

File: src/pages/blog/index.tsx [hunk 13]
```diff
@@ -122,25 +122,30 @@ const BlogPage: FC<{ data: BlogGQL }> = ({ data }) => {
...
   const developerProfile = useDeveloperProfile();
   const [selectedBlogpost, setSelectedBlogpost] = useState(-1);
+  const dateFormat = 'DD MMMM YYYY';
+
+  if (typeof window !== `undefined` && blogData.blogPost?.length === 0) {
+    navigate('/');
+  }
 
   // no blogpost will have index equal to -1 therefore no blogpost will be selected
   const handleCloseBlogpost = () => {
```
Comment (dominikslawkowski) [inline src/pages/blog/index.tsx]: typo here
```suggestion
  const handleCloseBlogPost = () => {
```

File: src/pages/blog/index.tsx [hunk 14]
```diff
@@ -122,25 +122,30 @@ const BlogPage: FC<{ data: BlogGQL }> = ({ data }) => {
   const { componentType, isDesktop } = useComponentType();
   const developerProfile = useDeveloperProfile();
   const [selectedBlogpost, setSelectedBlogpost] = useState(-1);
```
Comment (dominikslawkowski) [inline src/pages/blog/index.tsx]:
```suggestion
  const [selectedBlogPost, setSelectedBlogPost] = useState(-1);
```

File: src/pages/blog/index.tsx [hunk 15]
```diff
@@ -122,25 +122,30 @@ const BlogPage: FC<{ data: BlogGQL }> = ({ data }) => {
...
+  const dateFormat = 'DD MMMM YYYY';
+
+  if (typeof window !== `undefined` && blogData.blogPost?.length === 0) {
```
Comment (dominikslawkowski) [inline src/pages/blog/index.tsx]: Is the window condition necessary here? If so please move it to utils and create a method eg. `canUseWindow`

File: src/pages/blog/index.tsx [hunk 16]
```diff
@@ -160,32 +165,34 @@ const BlogPage: FC<{ data: BlogGQL }> = ({ data }) => {
...
-                    tagtitle={blogPost.blogLabel}
-                    title={blogPost.blogTitle}
-                    type={componentType}
-                    handleClose={() => handleCloseBlogpost()}
-                    handlePrev={() => handlePrevBlogpost(index)}
-                    handleNext={() => handleNextBlogpost(index)}
-                  />
-                </div>
-              ))}
+              {blogData.blogPost !== undefined
```
Comment (dominikslawkowski) [inline src/pages/blog/index.tsx]: Isn't it enough to use `!blogData.blogPost`?

Comment (dominikslawkowski) [reply src/pages/blog/index.tsx]: Also please use the `&&`, if you are not using the second return of the ternary operator

File: src/pages/blog/index.tsx [hunk 17]
```diff
@@ -160,32 +165,34 @@ const BlogPage: FC<{ data: BlogGQL }> = ({ data }) => {
...
+                      <BlogPostDialog
+                        contentheader={blogPost.blogDescription}
+                        contentmain={blogPost.blogBody}
+                        imgurl={blogPost.blogImage.publicURL}
+                        isOpen={index === selectedBlogpost}
+                        subtitle={dayjs(blogPost.publishDate).format(dateFormat)}
+                        tagtitle={blogPost.blogLabel}
+                        title={blogPost.blogTitle}
+                        type={componentType}
+                        handleClose={() => handleCloseBlogpost()}
```
Comment (dominikslawkowski) [inline src/pages/blog/index.tsx]:
```suggestion
                        handleClose={handleCloseBlogpost}
```

File: src/pages/index.tsx [hunk 18]
```diff
@@ -175,50 +180,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
-                <Testimonial
-                  key={`${item.testimonialName}`}
-                  isMobile={isMobile}
-                  image={item.testimonialImage.publicURL}
-                  description={item.testimonialText}
-                  labelBold={item.testimonialName}
-                  label={item.testimonialJob}
-                />
-              ))}
+              {hasTechnologies ? (
```
Comment (dominikslawkowski) [inline src/pages/index.tsx:183]: Please use the `&&`, if you are not using the second return of the ternary operator

File: src/pages/index.tsx [hunk 19]
```diff
@@ -175,50 +180,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
+                  <Box className={classes.skills}>
+                    {aboutData.socialMedia.technologies?.map((item) => (
+                      <Skill key={item.technologyName} level={item.technologyValue}>
+                        {item.technologyName}
+                      </Skill>
+                    ))}
+                  </Box>
+                </>
+              ) : null}
+              {hasTools ? (
```
Comment (dominikslawkowski) [inline src/pages/index.tsx:197]: Same above

File: src/pages/index.tsx [hunk 20]
```diff
@@ -175,50 +180,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
+                  <Box className={classes.skills}>
+                    {aboutData.socialMedia.tools?.map((item) => (
+                      <Skill key={item.toolName} level={item.toolValue}>
+                        {item.toolName}
+                      </Skill>
+                    ))}
+                  </Box>
+                </>
+              ) : null}
+              {hasOtherSkills ? (
```
Comment (dominikslawkowski) [inline src/pages/index.tsx:211]: Same above

File: src/pages/index.tsx [hunk 21]
```diff
@@ -175,50 +180,66 @@ const About: FC<{ data: AboutPageData }> = ({ data }) => {
...
+                    {aboutData.socialMedia.otherSkills?.map((item) => (
+                      <Skill key={item.otherSkillName} level={item.otherSkillValue}>
+                        {item.otherSkillName}
+                      </Skill>
+                    ))}
+                  </Box>
+                </>
+              ) : null}
             </Box>
+            {hasTestimonials ? (
```
Comment (dominikslawkowski) [inline src/pages/index.tsx:226]: Same above

**Review by dominikslawkowski — CHANGES_REQUESTED** [review]
Really good job 🏅 few comments

File: src/pages/blog/index.tsx (hunk 15 above)
Comment (mecies) [reply src/pages/blog/index.tsx]: It is, will move to utils 😄 
Explaination here => https://github.com/CodersCrew/CodersCard/pull/93#discussion_r504917067

File: src/pages/blog/index.tsx (hunk 16 above)
Comment (mecies) [reply src/pages/blog/index.tsx]: It should be enough, I was advised not to use `&&` and to switch to ternary above in this PR
Article here => https://github.com/CodersCrew/CodersCard/pull/93#discussion_r504246587

File: src/pages/index.tsx (hunk 18 above)
Comment (mecies) [reply src/pages/index.tsx:183]: reply here https://github.com/CodersCrew/CodersCard/pull/93#discussion_r505770592

File: src/pages/index.tsx (hunk 19 above)
Comment (mecies) [reply src/pages/index.tsx:197]: reply here https://github.com/CodersCrew/CodersCard/pull/93#discussion_r505770592

File: src/pages/index.tsx (hunk 20 above)
Comment (mecies) [reply src/pages/index.tsx:211]: reply here https://github.com/CodersCrew/CodersCard/pull/93#discussion_r505770592

File: src/pages/index.tsx (hunk 21 above)
Comment (mecies) [reply src/pages/index.tsx:226]: reply here https://github.com/CodersCrew/CodersCard/pull/93#discussion_r505770592

File: src/pages/blog/index.tsx (hunk 13 above)
Comment (mecies) [reply src/pages/blog/index.tsx]: Changed 😄

**Review by dominikslawkowski — APPROVED** [review]
awesome!
